source/common/network/apple_dns_impl.cc

```diff
@@ -0,0 +1,259 @@
+  //   reference's socket.
+  sd_ref_event_.reset();
+  DNSServiceRefDeallocate(main_sd_ref_);
+}
+
+void AppleDnsResolverImpl::initializeMainSdRef() {
+  auto error = DNSServiceCreateConnection(&main_sd_ref_);
+  if (error) {
+    throw EnvoyException("Error in DNSServiceCreateConnection");
+  }
+
+  auto fd = DNSServiceRefSockFD(main_sd_ref_);
+  if (fd == -1) {
+    throw EnvoyException("Error in DNSServiceRefSockFD");
+  }
+  ENVOY_LOG(warn, "DNS resolver has fd={}", fd);
+
+  sd_ref_event_ = dispatcher_.createFileEvent(
+      fd, [this](uint32_t events) { onEventCallback(events); }, Event::FileTriggerType::Level,
+      Event::FileReadyType::Read);
+  sd_ref_event_->setEnabled(Event::FileReadyType::Read);
+}
+
+void AppleDnsResolverImpl::onEventCallback(uint32_t events) {
+  ENVOY_LOG(warn, "DNS resolver file event");
+  if (events & Event::FileReadyType::Read) {
+    DNSServiceProcessResult(main_sd_ref_);
+  }
+}
+
+ActiveDnsQuery* AppleDnsResolverImpl::resolve(const std::string& dns_name,
+                                              DnsLookupFamily dns_lookup_family,
+                                              ResolveCb callback) {
+  ENVOY_LOG(warn, "DNS resolver resolve={}", dns_name);
+  std::unique_ptr<PendingResolution> pending_resolution(
+      new PendingResolution(*this, callback, dispatcher_, main_sd_ref_, dns_name));
+
+  DNSServiceErrorType error = pending_resolution->dnsServiceGetAddrInfo(dns_lookup_family);
+  if (error != kDNSServiceErr_NoError) {
+    return nullptr;
+  }
+
+  // If the query was synchronously resolved, there is no need to return the query.
+  // FIXME: I don't think synchronous resolution is possible from my read of dns_sd.h,
+  // and tests with local resolution.
+  // What about with caching?
+  if (pending_resolution->synchronously_completed_) {
+    return nullptr;
+  }
+
+  pending_resolution->owned_ = true;
+  return pending_resolution.release();
+}
+
+void AppleDnsResolverImpl::flushPendingQueries() {
+  ENVOY_LOG(warn, "DNS Resolver flushing {} queries", queries_with_pending_cb_.size());
+  for (std::list<PendingResolution*>::iterator it = queries_with_pending_cb_.begin();
+       it != queries_with_pending_cb_.end(); ++it) {
+    try {
+      ASSERT((*it)->pending_cb_);
+      (*it)->callback_((*it)->pending_cb_->status_, std::move((*it)->pending_cb_->responses_));
+    } catch (const EnvoyException& e) {
+      ENVOY_LOG(critical, "EnvoyException in DNSService callback: {}", e.what());
+      (*it)->dispatcher_.post([s = std::string(e.what())] { throw EnvoyException(s); });
+    } catch (const std::exception& e) {
+      ENVOY_LOG(critical, "std::exception in DNSService callback: {}", e.what());
+      (*it)->dispatcher_.post([s = std::string(e.what())] { throw EnvoyException(s); });
+    } catch (...) {
+      ENVOY_LOG(critical, "Unknown exception in DNSService callback");
+      (*it)->dispatcher_.post([] { throw EnvoyException("unknown"); });
+    }
```

Comment:
It this the same logic we have in the c-ares resolver? I didn't realize we are catching all exceptions like this. Seems not great. I wonder if this code could be shared between both resolvers?

Reply:
Yep, this is shared from the c-ares implementation. I could not find anything about exception safety inside the Apple library so I followed @htuch's example and took the exception outside the context of the callback.